vike/shared/page-configs/Config.ts

```diff
@@ -56,6 +67,18 @@ type HookReturnSync<Name extends keyof Config> = Exclude<HookReturnType<Name>, P
 type HookReturnAsync<Name extends keyof Config> = Extract<HookReturnType<Name>, Promise<any>>
 type HookReturn<Name extends keyof Config> = HookReturnAsync<Name>
 
+// NOTE(aurelien): to be extended. I'm wondering if this is all worth it as I'm wondering if people
+// will/should really use this style.
+type OnBeforeRenderParam = PageContextServer
+// NOTE(aurelien): unfortunately the user can't use directly `OnBeforeRenderReturnType` but needs
+// to explicitly select either the Async or the Sync version.
+type OnBeforeRenderReturnType = OptionalPromise<{ pageContext: Partial<Vike.PageContext> } | void>
+type OnBeforeRenderReturnAsync = Extract<OnBeforeRenderReturnType, Promise<any>>
+type OnBeforeRenderReturnSync = Exclude<OnBeforeRenderReturnType, Promise<any>>
+// NOTE(aurelien): but as discussed we still provide a simpler name for Async as default.
+type OnBeforeRenderReturn = OnBeforeRenderReturnAsync
```

Comment:
In order not to have a cryptic direct definition of `OnBeforeRenderReturn` for the user, maybe I should write instead something like this:

```ts
type OnBeforeRenderReturnType = OptionalPromise<{ pageContext: Partial<Vike.PageContext> } | void>
type OnBeforeRenderReturnAsync = Promise<{ pageContext: Partial<Vike.PageContext> } | void>
type OnBeforeRenderReturnSync = { pageContext: Partial<Vike.PageContext> } | void
type OnBeforeRenderReturn = Promise<{ pageContext: Partial<Vike.PageContext> } | void>
```

Reply:
I had the same idea :)